src/core/qgis.cpp

```diff
@@ -126,52 +126,115 @@ void qgsFree( void *ptr )
   free( ptr );
 }
 
-bool qgsVariantLessThan( const QVariant &lhs, const QVariant &rhs )
+int qgsVariantCompare( const QVariant &lhs, const QVariant &rhs )
 {
   // invalid < NULL < any value
   if ( !lhs.isValid() )
-    return rhs.isValid();
+  {
+    return rhs.isValid() ? -1 : 0;
+  }
   else if ( lhs.isNull() )
-    return rhs.isValid() && !rhs.isNull();
+  {
+    if ( !rhs.isValid() )
+      return 1;
+    if ( rhs.isNull() )
+      return 0;
+    return -1;
+  }
   else if ( !rhs.isValid() || rhs.isNull() )
-    return false;
+  {
+    return 1;
+  }
 
+  // both valid
   switch ( lhs.userType() )
   {
     case QMetaType::Type::Int:
-      return lhs.toInt() < rhs.toInt();
+    case QMetaType::Type::Char:
+    case QMetaType::Type::Short:
+    {
+      const int lhsInt = lhs.toInt();
+      const int rhsInt = rhs.toInt();
+      return lhsInt < rhsInt ? -1 : ( lhsInt == rhsInt ? 0 : 1 );
+    }
     case QMetaType::Type::UInt:
-      return lhs.toUInt() < rhs.toUInt();
+    case QMetaType::Type::UChar:
+    case QMetaType::Type::UShort:
+    {
+      const uint lhsUInt = lhs.toUInt();
+      const uint rhsUInt = rhs.toUInt();
+      return lhsUInt < rhsUInt ? -1 : ( lhsUInt == rhsUInt ? 0 : 1 );
+    }
     case QMetaType::Type::LongLong:
-      return lhs.toLongLong() < rhs.toLongLong();
+    case QMetaType::Type::Long:
+    {
+      const qlonglong lhsLongLong = lhs.toLongLong();
+      const qlonglong rhsLongLong = rhs.toLongLong();
+      return lhsLongLong < rhsLongLong ? -1 : ( lhsLongLong == rhsLongLong ? 0 : 1 );
+    }
     case QMetaType::Type::ULongLong:
-      return lhs.toULongLong() < rhs.toULongLong();
+    case QMetaType::Type::ULong:
+    {
+      const qulonglong lhsULongLong = lhs.toULongLong();
+      const qulonglong rhsULongLong = rhs.toULongLong();
+      return lhsULongLong < rhsULongLong ? -1 : ( lhsULongLong == rhsULongLong ? 0 : 1 );
+    }
     case QMetaType::Type::Double:
-      return lhs.toDouble() < rhs.toDouble();
+    {
+      const double lhsDouble = lhs.toDouble();
+      const double rhsDouble = rhs.toDouble();
+      return lhsDouble < rhsDouble ? -1 : ( lhsDouble == rhsDouble ? 0 : 1 );
```

Comment:
what about NaN comparisons ? (also true for Float).

I'd suggest considering NaN < any non-NaN value.

So 
```suggestion
      if ( std::isnan(lhsDouble) )
      {
           return std::isnan(rhsDouble) ? 0 : -1;
      }
      else if ( std::isnan(rhsDouble) )
      {
          return 1;
      }
      return lhsDouble < rhsDouble ? -1 : ( lhsDouble == rhsDouble ? 0 : 1 );
```

Reply:
Ok, added